### ISSUE TITLE
Provide command-line interface, or other API, for making a backup of a given grain

### ISSUE THREAD

**Issue by paulproteus** [sandstorm-io/sandstorm#1647]
### User story, cucumber-style

As a self-hosted Sandstorm user,

If I have an active grain that I care a great deal about,

And if I have substantial sysadmin knowledge,

Then I might want to use an automated system to make sure I take backups of that important grain on a schedule I control,

So I want Sandstorm to support an API (e.g. command line interface) to let me make a backup of that grain.

### Other notes

- I imagine we'd implement this as a `sudo sandstorm pause-grain {{grainId}}` command, then ask the user to make their snapshot, then `sudo sandstorm resume-grain {{grainId}}` command, but I'm just guessing.

- We're probably not going to be able to do this _super_ soon, but I wanted to leave a note here, findable by a search for "backups".

**Comment by ocdtrekkie** [comment]
I feel like this would be "better" if it was some sort of web interface I could target. If I could request a URL that would return my grain backup, that'd work on a local machine, but also for me to backup grains off Oasis, with the appropriate token.

**Comment by kpreid** [comment]
My use case: My personal laptop is more trusted and less internet-accessible than my Sandstorm server. Therefore, I should be able to _pull_ backups. Doing it over HTTP is a possibility; doing it over SSH is too, but that requires either passwordless sudo configuration or some other magic.

I also would like there to be a way to backup _all_ grains automatically. For example, if a grain backup is available at a URL as @ocdtrekkie suggested, then there can be a URL which yields a listing of all grains, so grabbing the full backup is a `wget -r`.

**Comment by ocdtrekkie** [comment]
Exactly. My biggest concern with the suggested `sandstorm backup grain` CLI type option is it limits the act of backing up grains by a script to a server admin function, when it's really a user function.

I have a shared hosting server where I can't access things like terminal, but I am able to generate database backup files, and I have a script on a home PC, which can open FTP connections, download all of the files, then cleanup the backup files on the server.

There's no point in implementing a only-on-the-server grain backup methodology, when one reachable through the Sandstorm server could also be scripted whether it's backing up on the same server or not. (Such as the aforementioned `wget`)

**Comment by paulproteus** [comment]
I'm leaving this comment to express my tentative +1 to emphasizing a
web-based API rather than a command-line-only API.

**Comment by zeigerpuppy** [comment]
how about using git?
Then the backups could be synchronised with the source, and it may be a good method to enable offline/online usage in the future.
For example, some people may prefer to work on a local LaTeX editor and then sync changes with a sandstorm LaTeX editor for when they're on the road...

**Comment by paulproteus** [comment]
I think practically speaking, we (Sandstorm core dev team) aren't going to
have time to implement a git interface for grain backups in the near
future. So our usual answer is: an app can support that.

But this is something that someone could make integrateable into every
other app. sandstorm-http-bridge is one example of something like that. In
the long run, something like sandstorm-http-bridge but that is, like,
sandstorm-grain-backup-git-synchronizer could be a binary that finds its
way into many apps, and even ends up part of vagrant-spk's default
generated package for apps.

So I just wanted to say that I like this git sync idea, but given how long
it sometimes takes the core team to do things (myself included), in general
the community moves faster if things can be pushed into apps.

**Comment by zenhack** [comment]
This could possibly be subsumed by (something like) the api is suggested at:

https://github.com/sandstorm-io/sandstorm/issues/2586#issuecomment-274180052

...plus a cli tool that talks to it.

**Comment by ocdtrekkie** [comment]
A user on IRC expressed a need for this today.

**Comment by ocdtrekkie** [comment]
This should be a lot easier with the new CLI feature, right?

**Comment by zenhack** [comment]
It would certainly be easier to call into sandstorm to do a backup from the command line. But per discussion above I think providing a way to do this via the UI is more important; you shouldn't have to be a server admin to do this, and I don't want key mechanisms like this to only be manageable by people with sysadmin knowledge.

In the issue I linked to about the API, I suggested offering this interface over a websocket, but I'm not sure that's really any easier than it was before; the js part of sandstorm could already provide that (though figuring out how via the meteor docs is something I had some trouble with the last time I looked at it).

**Comment by ocdtrekkie** [comment]
I think I agree with everything here, but that we probably should make sure our backup controls can also be triggered via the CLI, such that you can script with them, and use them in pre/post-backup features on conventional backup software. If I backup all of my machines with a given solution, I'd ideally perhaps want to have Sandstorm verify it has taken backups of everything before having my backup software backup where Sandstorm stores it's backups.

**Comment by zenhack** [comment]
Ideally what I'd like to do is have most functionality available via a capnproto API on some interface somewhere, and make sure we're using the same public API for most of what the UI does; this makes it hard for us to "cheat" and build UI functionality that isn't available programmatically.

Ultimately where I'd like to go with this is to have a "batteries included" backup solution for Sandstorm, so that users don't have to spend that much effort figuring out how to do backups; feed it some storage and it does the rest.

But it sounds like you already have some existing backup infra you want to integrate with? I'd like to get a more concrete picture of that. I don't want to invalidate important use cases, but I want to be moving us in the direction of "you don't need to be a sysadmin to get backups working." That would kinda suck.

**Comment by ocdtrekkie** [comment]
I try to think of these issues with regards to both environments I work with: home and work. (I only have a Sandstorm server at the former, but I would hope someday Sandstorm is a compelling case for use at the latter.)

At home I generally sync my backups between some local hard drives, and then off to a NAS which is generally living somewhere off-site. Pretty much everything assumes I'm working with files on an SMB share, so I am pretty happy if my server pastes it's backups into a folder that I can enable SMB on. (I think I tried setting up an NFS share on my NAS once and it didn't work... SMB on Linux was easier.)

At work we use an enterprise backup solution and a storage appliance. Some applications we host enjoy employing their own backup solutions, and usually that just makes it harder to back them up. We generally want to disable those as much as possible, and then employ our own backups of the entire virtual machines. But if I needed to stop the service or ensure it had stashed the backups safely prior to initiating the backup, that might be a thing I'd want to do.

**Comment by zenhack** [comment]
For the latter case, I wonder if a better approach would be to just install on top of a filesystem that supports snapshots (e.g. btrfs), and skip formal backups entirely; just snapshot /opt/sandstorm and be done with it.

**Comment by ocdtrekkie** [comment]
Entirely possible, although as a non-Linux-familiar person, I have no idea what btrfs is and whether or not I have it. ;)

**Comment by zenhack** [comment]
https://en.wikipedia.org/wiki/Btrfs and in particular https://en.wikipedia.org/wiki/Btrfs#Subvolumes_and_snapshots

You'd have to set it up when formatting the drive, but it's been around long enough at this point that it's going to be on option on most any distro.

**Comment by ocdtrekkie** [comment]
I guess when you come from Windows, it's weird to have a default file system that doesn't support snapshots. But yeah, I suppose I'd use that in an enterprise environment then.

**Comment by rasos** [comment]
As sysadmin I would like to be able to downlowad a backup of any grain. 

We have the case that users of a sandpass grain would like to download a backup of that grain, but non of them can see the download arrow in the Web GUI and the original creator of the user has no access anymore to the sandstorm instance. So I suggested they share that grain with me as an admin, but even me can not see the download arrow.

**Comment by ocdtrekkie** [comment]
@rasos In most cases, admins do not have any special powers over individual grains within the UI. I think ideally admins should gave a feature to reassign ownership of a grain.

What admins can do, arguably, right now, is access the file system Sandstorm is installed in. What's in the appropriate directory inside the grains directory is essentially the contents of the grain backup, minus the metadata, which you can recreate by downloading an empty Sandpass grain.

**Comment by rasos** [comment]
@ocdtrekkie ah I did not know that accessing a grain is that easy.

Just for the records for anybody else who may need that as well: in our installation all grains are simply at `/opt/sandstorm/var/sandstorm/grains/` with the hash being the same like in the grains URL. 

Yes agree, a feature to re-assign ownership of a grain may be useful.

**Comment by ocdtrekkie** [comment]
Glad you found it. Definitely a goal for Sandstorm is to be able to support any kind of need you have without you needing to dig into either the file system or the database, but it's a lot faster than waiting for a feature development. :)

**Comment by zenhack** [comment]
Fwiw, i see no reason not to allow backups for admin users of any grain they can see. It would be a pretty simple change.

Once the grain settings page has landed, we can start putting stuff re: ownership management in there.
